clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +262,465 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+                return;
+            }
+
+            // Partitions to assign (not previously owned)
+            SortedSet<TopicPartition> addedPartitions = new TreeSet<>(COMPARATOR);
+            addedPartitions.addAll(assignedPartitions);
+            addedPartitions.removeAll(ownedPartitions);
+
+
+            // Partitions to revoke
+            SortedSet<TopicPartition> revokedPartitions = new TreeSet<>(COMPARATOR);
+            revokedPartitions.addAll(ownedPartitions);
+            revokedPartitions.removeAll(assignedPartitions);
+
+            log.info("Updating assignment with\n" +
+                            "\tAssigned partitions:                       {}\n" +
+                            "\tCurrent owned partitions:                  {}\n" +
+                            "\tAdded partitions (assigned - owned):       {}\n" +
+                            "\tRevoked partitions (owned - assigned):     {}\n",
+                    assignedPartitions,
+                    ownedPartitions,
+                    addedPartitions,
+                    revokedPartitions
+            );
+
+            CompletableFuture<Void> revocationResult;
+            if (!revokedPartitions.isEmpty()) {
+                revocationResult = revokePartitions(revokedPartitions);
+            } else {
+                revocationResult = CompletableFuture.completedFuture(null);
+                // Reschedule the auto commit starting from now (new assignment received without any
+                // revocation).
+                commitRequestManager.resetAutoCommitTimer();
+            }
+
+            // Future that will complete when the full reconciliation process completes (revocation
+            // and assignment, executed sequentially)
+            CompletableFuture<Void> reconciliationResult =
+                    revocationResult.thenCompose(r -> {
+                        if (state == MemberState.RECONCILING) {
+                            // Make assignment effective on the client by updating the subscription state.
+                            subscriptions.assignFromSubscribed(assignedPartitions);
+                            // Invoke user call back
+                            return invokeOnPartitionsAssignedCallback(addedPartitions);
+                        } else {
+                            // Revocation callback completed but member already moved out of the
+                            // reconciling state.
+                            CompletableFuture<Void> res = new CompletableFuture<>();
+                            res.completeExceptionally(new KafkaException("Interrupting " +
+                                    "reconciliation after revocation, as the member already " +
+                                    "transitioned out of the reconciling state into " + state));
+                            return res;
+                        }
+                    });
+
+            reconciliationResult.whenComplete((result, error) -> {
```

Comment:
I believe the reconciliation result is completed by the main thread.

Reply:
Just for the record, as discussed offline, none of these futures are expected to be completed in the main thread. This is a reconciliation future, which is much more than the callbacks (metadata, commit, callbacks). And callbacks, when implemented, will be based on events shared between the App thread and the Background thread (not based on futures from one thread complete on the other, as that would be problematic)